### ISSUE TITLE
W3CDom.fromJsoup fails when xmlns is defined

### ISSUE THREAD

**Issue by dportabella** [jhy/jsoup#1096]
```
import org.jsoup.Jsoup;
import org.jsoup.helper.W3CDom;
import org.jsoup.nodes.Document;
import org.w3c.dom.NodeList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpression;
import javax.xml.xpath.XPathFactory;
import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;

public class Test130 {
    public static void main(String[] args) throws Exception {
        String html = "<html xmlns=\"http://www.w3.org/1999/xhtml\"><body><div>hello</div></body></html>";  // this input fails with jsoup W3CDom
        // String html = "<html><body><div>hello</div></body></html>";                                      // this input works with jsoup W3CDom

        org.jsoup.nodes.Document jsoupDoc = Jsoup.parse(html);

        test(new W3CDom().fromJsoup(jsoupDoc));            // returns 0, incorrect
        test(toW3CDocWorkaround(jsoupDoc));                // returns 1, correct
    }

    private static void test(org.w3c.dom.Document w3cDoc) throws Exception {
        XPathExpression xpath = XPathFactory.newInstance().newXPath().compile("/html");
        System.out.println(((NodeList) xpath.evaluate(w3cDoc, XPathConstants.NODESET)).getLength());
    }

    private static org.w3c.dom.Document toW3CDocWorkaround(org.jsoup.nodes.Document jsoupDoc) throws Exception {
        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
        return documentBuilder.parse(new ByteArrayInputStream(jsoupDoc.toString().getBytes(StandardCharsets.UTF_8)));
    }
}
```

This seems related to https://github.com/jhy/jsoup/issues/671, but it's not the same issue.

**Comment by panthony** [comment]
I got the exact same issue, when `xmls` is set I cannot apply XPathExpression anymore it doesn't match anything.

The `toW3CDocWorkaround` doesn't work for me, first test case I got the following error after ~30s stuck in `documentBuilder.parse`:

> [Fatal Error] :43:17: Attribute name "async" associated with an element type "script" must be followed by the ' = ' character.
org.xml.sax.SAXParseException; lineNumber: 43; columnNumber: 17; Attribute name "async" associated with an element type "script" must be followed by the ' = ' character.

Any pointer on how this could be fixed would be appreciated 👍

**Comment by dportabella** [comment]
@panthony, you mean the first test case here (where there is an `async` attribute)? https://github.com/jhy/jsoup/issues/1098

Try the `w3cDocToString` method proposed on https://github.com/jhy/jsoup/issues/1098.

**Comment by dportabella** [comment]
glad to hear that `w3cDocToString` works as a workaround. I hope this issue will be fixed also.
For the other issue, I guess it's better that you open a new ticket if necessary.

**Comment by SimonSchmid** [comment]
Hello all,

I think I have the same issue as @panthony . I am trying to parse an html which contains an svg that has xmlns already defined. I am converting the parsed document to an `org.w3c.dom.Document` with
`org.w3c.dom.Document w3cDoc = new W3CDom().fromJsoup(htmlDocument);`
When I try to parse the string representation of this `w3cDoc` with `javax.xml.parsers.SAXParser`, I get the following exception:

> org.xml.sax.SAXParseException; lineNumber: 965; columnNumber: 218; Attribute "xmlns" bound to namespace "http://www.w3.org/2000/xmlns/" was already specified for element "svg".

My problem is that I have some intermediate steps, so that I need to have w3c document, i.e., the `w3cDocToString`fix does not work for me and I did not find any working way to convert the string coming out of `w3cDocToString` back to a w3c document. If am trying to parse this string with the `SAXParser`, I get:

> org.xml.sax.SAXParseException; lineNumber: 16; columnNumber: 32; The entity "copy" was referenced, but not declared.

The website I am trying to parse is this one btw: http://www.handletheheat.com/peanut-butter-snickerdoodles 

My hacky fix is currently to remove the xlmns attribute before convertig the JSoup document to w3c. However, this leads to a missing xmlns attribute.

Besides this issue, JSoup does work really nice for me and I appreciate all the work you have done!

Thank you for any help,
Simon

**Comment by garretwilson** [comment]
What's the status of this bug? I am writing a a static site generator which already has XHTML and Markdown support, but I want to add HTML support. jsoup is getting good reviews, but real W3C DOM support is essential; we can't break if `xmlns` isn't handled correctly.

Are there other W3C DOM implementation problems I should be aware of before we start trying to integrate jsoup? I don't want to spend weeks only to find out that it doesn't work with a "real" DOM. Thanks.

**Comment by jhy** [comment]
This actually works OK AFAICT. The W3C Document emitted by Jsoup W3C is namespace aware. So you need to make your xpath query namespaced as well. See e.g. https://stackoverflow.com/a/4440519/153184

fe92494c7716eeb04d5d4040cc43d67b8f2f4aaf fixes the XML string output
052dc1b1cac3dd4bfded785ce8c9dadc8b95f18c verifies the xpath selectors with / without namespaces

(And the build is broken right now because indentation default is specified by the platform and different on Travis than my local, will fix)

@SimonSchmid if you keep seeing that issue in HEAD, can you file an issue with a complete code sample please.
